Update `pg-backup-api` docs to cover all supported operations (#5169)
EnterpriseDB/docs: advocacy_docs/supported-open-source/barman/pg-backup-api/03-tasks.mdx

`OPERATION_ID` is a short string that represents the ID of an operation created by the API.

You can find out how to fetch operations created by the API below

### Available operations

#### GET: /operations

Returns all instance level tasks created by the API, if any.

```bash
curl http://barman-server.my.org/operations
{
  "operations": [
    {
      "id": "20240118T104231",
      "type": "config_update"
    },
    {
      "id": "20240118T103947",
      "type": "config_update"
    }
  ]
}
```

#### POST: /operations

Request pg-backup-api to create instance level operations.

Currently pg-backup-api supports only the `config_update` operation.

!!! Note
    You need to set the `content-type` header as `application/json` when executing `POST` requests to this endpoint, otherwise you will receive a `400 Bad Request error`.

##### config_update

This operation is used to request a remote execution of `barman config-update` command.

The JSON payload of the request must contain the following keys:

 * `type`: set as `config_update` to request that operation
 * `changes`: an array of documents to be used as argument for `barman config-update`

> As an example, you could request pg-backup-api to update the configuration of a couple Barman servers plus a Barman model with a payload like this:

Reviewer: ```suggestion
As an example, you could request pg-backup-api to update the configuration of a couple of Barman servers plus a Barman model with a payload like this:
```

```json
{
  "type": "config_update",
  "changes": [
    {
      "scope": "server",
      "server_name": "my-server-a",
      "archiver": "on",
      "streaming_archiver": "off"
    },
    {
      "scope": "server",
      "server_name": "my-other-server-b",
      "conninfo": "some_conn_info",
      "streaming_conninfo": "some_streaming_conninfo",
      "streaming_archiver": "on"
    },
    {
      "scope": "model",
      "model_name": "some-model",
      "streaming_conninfo": "some_streaming_conn_info"
    }
  ]
}
```

> Assume the above JSON content is stored in a file named `payload-pg-backup-api.json`, you could issue a request to the API like this:

Reviewer: ```suggestion
Assume the above JSON content is stored in a file named `payload-pg-backup-api.json`, you could issue a request to the API using `curl` in the following way:
```

```bash

> curl -X POST http://barman-server.my.org/operations -H "content-type: application/json" -d@payload-pg-backup-api.json

Reviewer: ```suggestion
curl -X POST http://barman-server.my.org/operations -H "content-type: application/json" -d @payload-pg-backup-api.json
```

{
  "operation_id": "20240118T104231"
}
```

The `operation_id` in the response is the ID of the operation that has been created in by the API.

#### GET: /operations/OPERATION_ID

This endpoint allows you to check the status of an instance level operation with ID `OPERATION_ID`, which can be either of these:

 * `DONE`: if the operation finished successfully
 * `FAILED`: if the operation finished with errors
 * `IN_PROGRESS`: if the operation is still ongoing

 Please find an example below:

 ```bash
 curl http://barman-server.my.org/operations/20240118T104231
 {
  "operation_id": "20240118T104231",
  "status": "DONE"
}
 ```

#### GET: /servers/NAME/operations

> Returns all server level tasks created by the API for Barman server `NAME`, if any.

Reviewer: Add below:
The following example shows the two tasks created on the `my-barman-server` Postgres server.

> Returns all server level tasks created by the API for Barman server `NAME`, if any.

Author: ```suggestion
Returns all server level tasks created by the API for Barman server `NAME`, if any.

The following example shows the two tasks created on the `my-barman-server` Barman server.
```

Currently pg-backup-api supports `recovery` and `config_switch` operations.

!!! Note
    You need to set the `content-type` header as `application/json` when executing `POST` requests to this endpoint, otherwise you will receive a `400 Bad Request error`.

##### recovery

This operation is used to request a remote execution of `barman recover` command.

The JSON payload of the request must contain the following keys:

 * `type`: set as `recovery` to request that operation;
 * `backup_id`: ID of the backup to be restored from Barman server `NAME`. Supports whatever `barman recover` supports, including `latest` for example;
 * `remote_ssh_command`: the SSH command to be used to connect to the target Postgres server where the backup will be restored;
 * `destination_directory`: the path in the target server where the backup will be restored.

As an example, you could request pg-backup-api to restore the latest backup of Barman server `db_server_two` with a payload like this:

```json
{
  "type": "recovery",
  "backup_id": "latest",
  "remote_ssh_command": "ssh postgres@db_server_two.my.org",
  "destination_directory": "/var/lib/pgdata"
}
```

```bash

> curl -X POST http://barman-server.my.org/servers/db_server_two/operations -H "content-type: application/json" -d@payload-pg-backup-api.json

Reviewer: ```suggestion
curl -X POST http://barman-server.my.org/servers/db_server_two/operations -H "content-type: application/json" -d @payload-pg-backup-api.json
```

{
  "operation_id": "20240118T140951"
}
```

The `operation_id` in the response is the ID of the operation that has been created in by the API.

##### config_switch

This operation is used to request a remote execution of `barman config-switch` command.

The JSON payload of the request must contain the following key:

 * `type`: set as `config_switch` to request that operation.

Besides that, it may contain either of the following keys:

 * `model_name`: the name of a model, if you want to apply it to the Barman server `NAME`; or
 * `reset`: set with `true` if you want to unapply the currently active model of the Barman server `NAME`

As an example, you could request pg-backup-api to apply the model `my-model` to the Barman server with a payload like this:

```json
{
  "type": "config_switch",
  "model_name": "my-model"
}
```

```bash

> curl -X POST http://barman-server.my.org/servers/db_server_two/operations -H "content-type: application/json" -d@payload-pg-backup-api.json

Reviewer: ```suggestion
curl -X POST http://barman-server.my.org/servers/db_server_two/operations -H "content-type: application/json" -d @payload-pg-backup-api.json
```

 ```bash
 curl http://barman-server.my.org/operations/20240118T104231
 {
  "operation_id": "20240118T104231",
  "status": "DONE"
}
 ```

#### GET: /servers/NAME/operations

Returns all server level tasks created by the API for Barman server `NAME`, if any.

The following example shows the two tasks created on the `my-barman-server` Barman server.

```bash
curl http://barman-server.my.org/servers/my-barman-server/operations
{
  "operations": [
    {
      "id": "20240118T140951",
      "type": "recovery"
    },
    {
      "id": "20240118T141135",
      "type": "config_switch"
    }
  ]
}
```

#### POST: /servers/NAME/operations

Request pg-backup-api to create server level operations.

Currently pg-backup-api supports `recovery` and `config_switch` operations.

!!! Note
    You need to set the `content-type` header as `application/json` when executing `POST` requests to this endpoint, otherwise you will receive a `400 Bad Request error`.

##### recovery

This operation is used to request a remote execution of `barman recover` command.

The JSON payload of the request must contain the following keys:

 * `type`: set as `recovery` to request that operation;

>  * `backup_id`: ID of the backup to be restored from Barman server `NAME`. Supports whatever `barman recover` supports, including `latest` for example;

Reviewer 2: That line "Supports whatever `barman recover` supports, including `latest` for example;"

I might re-phrase it into
You could use Barman shortcuts or anything that `barman recover` support, including `latest` for example;

>  * `backup_id`: ID of the backup to be restored from Barman server `NAME`. Supports whatever `barman recover` supports, including `latest` for example;

Author: ```suggestion
 * `backup_id`: ID of the backup to be restored from Barman server `NAME`. You could use Barman shortcuts or anything that `barman recover` supports, including `latest` for example;
```

}
```

```bash
curl -X POST http://barman-server.my.org/servers/db_server_two/operations -H "content-type: application/json" -d @payload-pg-backup-api.json
{
  "operation_id": "20240118T140951"
}
```

The `operation_id` in the response is the ID of the operation that has been created in by the API.

##### config_switch

This operation is used to request a remote execution of `barman config-switch` command.

The JSON payload of the request must contain the following key:

 * `type`: set as `config_switch` to request that operation.

Besides that, it may contain either of the following keys:

 * `model_name`: the name of a model, if you want to apply it to the Barman server `NAME`; or
 * `reset`: set with `true` if you want to unapply the currently active model of the Barman server `NAME`

As an example, you could request pg-backup-api to apply the model `my-model` to the Barman server with a payload like this:

```json
{
  "type": "config_switch",
  "model_name": "my-model"
}
```

```bash
curl -X POST http://barman-server.my.org/servers/db_server_two/operations -H "content-type: application/json" -d @payload-pg-backup-api.json
{
  "operation_id": "20240118T141135"
}
```

> The `operation_id` in the response is the ID of the operation that has been created in by the API.

Reviewer 2: I think we need a comma "The `operation_id` in the **response,** is the ID "

Author: I would use a comma if the sentence was like this:

"The operation_id in the response, `it` is the ID"

I'm not sure if a comma applies without an `it` or anything else to take a "break" in the sentence.

What do you think?

And what do you think @martinmarques ? You are way better than me on formal writing 😆

Reviewer 2: Yeah, we don't need "it" in this case because the phrase is focusing on describing the ID

Reviewer: We should try to avoid using "it" as much as possible

Reviewer: No need for a coma there

#### GET: /servers/NAME/operations/OPERATION_ID

> This endpoint allows you to check the status of an server level operation for server `NAME` with ID `OPERATION_ID`, which can be either of these:

Reviewer 2: "an server" should be "a server"

> This endpoint allows you to check the status of an server level operation for server `NAME` with ID `OPERATION_ID`, which can be either of these:

Author: ```suggestion
This endpoint allows you to check the status of a server level operation for server `NAME` with ID `OPERATION_ID`, which can be either of these:
```
